[TransformStrategies] Add support for aligned and partially aligned matmul

qedawkins opened the pull request:
Current handling of tile aligned matmuls differs from unaligned in the way the tensor.pad ops are lowered. When the pad is aligned and would otherwise fold, we convert to a linalg.copy and don't need to do masked vectorization. In the future the plan is to unify aligned and unaligned cases.

Currently due to some observed performance regressions, any alignment is guarded behind `--iree-codegen-llvmgpu-enable-transform-dialect-aligned-matmul`.

qedawkins committed:
[TransformStrategies] Enable aligned and partially aligned matmul

Current handling of tile aligned matmuls differs from unaligned in the
way the tensor.pad ops are lowered. When the pad is aligned and would
otherwise fold, we convert to a linalg.copy and don't need to do masked
vectorization. In the future the plan is to unify aligned and unaligned
cases.

nicolasvasilache commented on compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:
```diff
@@ -839,23 +868,6 @@ static LogicalResult matchAndSetMatmulStrategy(func::FuncOp entryPoint,
-  //   - n and k are not aligned to the tile sizes (conservatively, take 64, 16)
-  // Other cases currently result in folding and fall back to the default
-  // unaligned IREE strategy.
-  bool supportedUnalignedCases =
```
There is a concurrent effort to activate the unaligned case by default.

Could we keep the separation here and introduce a new flag (off by default) to activate the aligned case ?

I'd like to keep the 2 concerns separate because I suspect there is some missing stuff in the aligned case vs the current IREE pipeline.

qedawkins replied:
Sounds good. Just trying a few benchmarks reveals that aligned here is a little behind default IREE (even picking the same tile sizes) so will add an additional flag.

qedawkins replied:
Partially aligned seems to still be better than the current default SIMT path so I left that out of the flag. (e.g. 650µs vs 2ms on A100 in the example in the tests with whatever tile sizes are picked by default).

Although my benchmarking was fairly ad-hoc so would be curious to know how others are doing it. I didn't see how to do it from the dispatch profiler at least.

nicolasvasilache commented on compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp:
```diff
@@ -285,12 +294,15 @@ void iree_compiler::gpu::buildMatmulTensorCoreStrategy(
-  // TODO: not a functional style op to avoid invalidating artificially.
-  funcH = b.create<transform::MaterializeMasksOp>(
-      pdl::OperationType::get(b.getContext()), funcH);
+  if (!strategy.alignedLhs() || !strategy.alignedRhs() ||
```
Are these guards actually necessary?
This should do nothing in case there are no masks right ?
If possible, I'd rather keep simplicity here.

nicolasvasilache commented on compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp:
```diff
@@ -235,7 +235,16 @@ void iree_compiler::gpu::buildMatmulTensorCoreStrategy(
+  } else {
+    fillOpH = b.create<transform::MatchOp>(variantH,
+                                           linalg::FillOp::getOperationName());
+    ApplyPatternsOpPatterns config;
```
Can you add a comment as to why this is necessary?
I'll try and remove it in a followup.

qedawkins replied:
(Leaving a longer comment here first)

First for rematching the fill, I see that I can just use the fill handle above from block distribution so that part is unnecessary and I will drop it with the rebase.

The important part here is running canonicalization. For aligned right now we are relying on canonicalization to simplify the `tensor.pad` such that once we rewrite in DPS we get a `linalg.copy`. For example, without canonicalization we will have something like:
```
%padded = tensor.pad %extracted_slice_8 nofold low[%c0_12, %c0_12] high[%14, %15] {
^bb0(%arg5: index, %arg6: index):
   tensor.yield %cst_11 : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
and after DPS
```
%17 = linalg.fill ins(%cst_16 : f16) outs(%16 : tensor<128x16xf16>) -> tensor<128x16xf16>
%inserted_slice = tensor.insert_slice %extracted_slice_8 into %17[%c0_12, %c0_12] [128, 16] [1, 1] : tensor<128x16xf16> into tensor<128x16xf16>
```
Whereas if we do canonicalization between the pad and DPS then we get
```
%padded = tensor.pad %extracted_slice_2 nofold low[%c0, %c0] high[%c0, %c0] {
^bb0(%arg5: index, %arg6: index):
  tensor.yield %cst : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
into
```
%11 = bufferization.alloc_tensor() : tensor<128x16xf16>
%12 = linalg.copy ins(%extracted_slice_2 : tensor<128x16xf16>) outs(%11 : tensor<128x16xf16>) -> tensor<128x16xf16>
```
(also annoyingly if we do canonicalization after rewriting to dps instead of before all the copies/pads get folded away)
```
%extracted_slice_2 = tensor.extract_slice %extracted_slice[0, %arg3] [128, 16] [1, 1] : tensor<128x1024xf16> to tensor<128x16xf16>
%extracted_slice_3 = tensor.extract_slice %extracted_slice_0[%arg3, 0] [16, 128] [1, 1] : tensor<1024x128xf16> to tensor<16x128xf16>
%11 = linalg.matmul ins(%extracted_slice_2, %extracted_slice_3 : tensor<128x16xf16>, tensor<16x128xf16>) outs(%arg4 : tensor<128x128xf16>) -> tensor<128x128xf16>
```

Right now, `HoistOutputPaddingOp` runs canonicalization as a last step, but we shouldn't do hoisting if the result is unaligned. Also IIRC if we canonicalize before hoisting then there is a chance that the extract_slice on the loop variable gets folded out causing hoisting to fail (definitely some brittleness to address there) so canonicalizing after the pad isn't an option. To summarize

1. If any pads are aligned then we must canonicalize before converting to DPS to get the copy
2. If the output is unaligned, we will run canonicalization anyway so aligned input pads will get the canonicalization they need anyway.
3. Canonicalization cannot run between padding and hoisting and we do not want to hoist an aligned output.

I can just make canonicalization unconditionally run after conditionally doing the output hoisting to remove the `else`, although I figured it would be ugly to run canonicalization back-to-back in the unaligned case. WDYT?

nicolasvasilache commented on compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:
```diff
@@ -453,24 +481,32 @@ void mlir::iree_compiler::gpu::buildMatmulVectorization(
       b.create<transform::MatchOp>(variantH, func::FuncOp::getOperationName());
   // TODO: avoid functional style transform so we can apply to the variant.
-  funcH = b.create<transform::LowerMaskedTransfersOp>(funcH.getType(), funcH);
+  if (!strategy.alignedLhs() || !strategy.alignedRhs() ||
```
can we also keep this unconditional?
Nothing should happen when unneeded

nicolasvasilache reviewed (approved):
Thanks @qedawkins and sorry for the delay, I had some local challenges that impacted my schedule.
Generally this looks good to me, this will require a few refactoring followups but it is great you were able to unify the 2 cases already.

Some minor comments around the fact that I'd rather reduce conditional complexity and let the transforms "do nothing" where applicable.

qedawkins committed:
Address comments and add tests for partially unaligned

This also adjusts the check for alignment to be based on the block tile
sizes instead of hard coded values.

qedawkins committed:
run tests

qedawkins committed:
Avoid canonicalization after tiling that can later break hoisting

qedawkins committed:
Turn off partially aligned
